File: customview/src/main/java/com/harry/customview/measure/MainActivity.java
package com.harry.customview.measure;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import com.harry.customview.R;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }
}

/**
 * 自定义View
 * 布局：onLayout onMeasure layout(ViewGroup)
 * 显示：onDraw
 * 交互：onTouchEvent
 * <p>
 * 执行流程：（搞清楚先后顺序，避免取值时机错误）
 * 1.构造函数（View初始化 2.onMeasure（测量View大小 3.onSizeChanged（确定View大小
 * 4.onLayout（确定子View布局 5.onDraw（绘制内容 6.视图状态改变会invalidate()并重新onDraw
 * <p>
 * 自定义ViewGroup最重要是实现onMeasure和onLayout
 * 自定义View最重要是实现onMeasure和onDraw
 * <p>
 * ViewPager设置wrap_content无效。因为ViewPager在onMeasure中，先setMeasuredDimension设置了自己的宽高，没有根据子View的尺寸
 * 网上博客也是错误的。因为在测量子View尺寸是都没有用到子View的LayoutParams，没有真正去测量子View的尺寸
 */

// TODO: 2022/3/11 作业：1.用Kotlin手写FlowLayout 2.学习FlexboxLayout 3.阅读FrameLayout和LinearLayout源码

File: coroutine/src/main/java/com/harry/coroutine/MainActivity.kt
package com.harry.coroutine

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity() {
    private val TAG = "MainActivity"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    // 用runBlocking将线程环境切换到协程环境。即有两个协程
    fun click1(view: android.view.View) = runBlocking {// 外协程
        // 默认是主线程，为launch指定参数即指定线程
        launch(Dispatchers.IO) {// 内协程
            Log.e(TAG, "click1: launch ${Thread.currentThread().name}")

            repeat(10) {
                Thread.sleep(1000)
                Log.d(TAG, "click1: 计数${it}")
            }
        }
    }

    // 协程的应用
    fun click2(view: android.view.View) = runBlocking {
        launch {
            // 这里是主线程

            // async可以返回线程的执行结果。默认也是主线程，通过参数指定为IO子线程
            val def = async(Dispatchers.IO) {
                // 这里是子线程
                Thread.sleep(1000)
                "模拟网络请求的结果"
            }

            // 这里是主线程
            // 用了高阶，所以async返回什么类型，await就得到什么类型
            tv_show.text = def.await()
        }


        // 简化版
        launch {
            tv_show.text = async(Dispatchers.IO) {
                Thread.sleep(1000)
                "模拟网络请求的结果"
            }.await()
        }
    }

    // 协程中频繁切换线程。对比RxJava的功能
    fun click3(view: android.view.View) = runBlocking {
        // 因为用的runBlocking，是阻塞式的

        Log.d(TAG, "click3: ${Thread.currentThread().name}")

        // 这里是主线程
        withContext(Dispatchers.IO) {
            // 这里是子线程
            Log.d(TAG, "click3: ${Thread.currentThread().name}")
        }

        // 这里是主线程
        Log.d(TAG, "click3: ${Thread.currentThread().name}")

        withContext(Dispatchers.IO) {
            // 这里是子线程
            Log.d(TAG, "click3: ${Thread.currentThread().name}")
        }

        // 这里是主线程
        Log.d(TAG, "click3: ${Thread.currentThread().name}")
    }


    // 与上面的区别到底是什么？？？
    // 非阻塞的方式，尽量不用阻塞式协程
    fun click4(view: android.view.View) {
        // 只用了一个协程
        GlobalScope.launch(Dispatchers.Main) {
            // 这里是主线程
            Log.d(TAG, "click3: ${Thread.currentThread().name}")

            withContext(Dispatchers.IO) {
                // 这里是子线程
                Log.d(TAG, "click3: ${Thread.currentThread().name}")
            }

            // 这里是主线程
            Log.d(TAG, "click3: ${Thread.currentThread().name}")

            withContext(Dispatchers.IO) {
                // 这里是子线程
                Log.d(TAG, "click3: ${Thread.currentThread().name}")
            }

            // 这里是主线程
            Log.d(TAG, "click3: ${Thread.currentThread().name}")
        }
    }


}

/*
线程：
操作系统来调度
依附于进程，对进程而言，线程是轻量级的
创建一万个线程，会崩溃，内存溢出
 */

/*
协程：
用户代码来控制
对线程而言，协程是轻量级的
创建一万个协程，没问题
 */

/**
 * 协程可参考“Kotlin中文社区”的简书文章
 * 比RxJava切换线程更方便
 *
 * 协程的原理视频？？？
 */

fun test01() {

    // GlobalScope.launch类似于守护线程，若主线程执行完毕，里面的子线程则不再执行
    GlobalScope.launch {
        delay(2000)
        println("1111")// 不会打印
    }

    println("A")
}

suspend fun test02() {
    val job = GlobalScope.launch {
        repeat(100) {
            delay(40)
            println(it)
        }
    }
    println("A")
    Thread.sleep(100)
    // 取消协程。时间不准确，会有一点时间差
    job.cancel()
    // 立刻取消协程。且函数要加上suspend
    job.cancelAndJoin()
}